flowmachine/flowmachine/features/location/majority_location.py

```diff
@@ -0,0 +1,75 @@
+from typing import List
+
+from flowmachine.core import Query
+from flowmachine.core.mixins import GeoDataMixin
+from flowmachine.features.utilities.subscriber_locations import BaseLocation
+
+
+class MajorityLocation(BaseLocation, Query):
+    """
+    A class for producing a list of subscribers along with the location (derived from `spatial_unit')
+    that they visited more than half the time. Takes a query that includes a 'subscribers' column,
+    a 'spatial_unit' attribute and a column to be used as weighting for locations (`location_count` for example)
+
+    Parameters
+    ----------
+    subscriber_location_weights: Query
+        The query object containing subscribers, locations, and weights.
+    weight_column: str
+        The column, when summed, that will produce the count used to threshold the majority
+    include_unlocatable: bool default False
+        If `True`, returns every unique subscriber in the `subscriber_location_weights` query, with
+        the location column as `NULL` if no majority is reached.
+        If `False`, returns only subscribers that have achieved a majority location
+    """
+
+    def __init__(
+        self,
+        subscriber_location_weights: Query,
+        weight_column: str,
+        include_unlocatable: bool = False,
+    ):
+        if "subscriber" not in subscriber_location_weights.column_names:
+            raise ValueError("`subscriber` not in subscriber_location_weights query")
+        if weight_column not in subscriber_location_weights.column_names:
+            raise ValueError("weight_column must exist in subscriber_subset")
+        if not getattr(subscriber_location_weights, "spatial_unit"):
+            raise AttributeError(
+                "subscriber_location_weights needs a spatial_unit column"
+            )
+
+        self.subscriber_location_weights = subscriber_location_weights
+        self.weight_column = weight_column
+        self.include_unlocatable = include_unlocatable
+        self.spatial_unit = subscriber_location_weights.spatial_unit
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        return ["subscriber"] + self.spatial_unit.location_id_columns
+
+    def _make_query(self):
+        loc_id = ",".join(self.spatial_unit.location_id_columns)
+        sql = f"""
+WITH subscriber_subset AS (
+    {self.subscriber_location_weights.get_query()}
+), summed_weights AS (
+    SELECT subscriber, sum({self.weight_column}) AS total_weight
+    FROM subscriber_subset
+    GROUP BY subscriber
+), seen_subs AS (
+    SELECT subscriber, {loc_id}
+    FROM summed_weights JOIN subscriber_subset USING(subscriber)
+    WHERE {self.weight_column} > total_weight/2.0
+)
+"""
+
+        if self.include_unlocatable:
+            sql += f"""
+SELECT subscriber, seen_subs.{loc_id}
+FROM seen_subs RIGHT OUTER JOIN summed_weights USING(subscriber)
+            """
+        else:
+            sql += f"""SELECT subscriber, {loc_id} FROM seen_subs"""
```

Comment:
Marginally nice to write this as "SELECT subscriber, seen_subs.{loc_id} FROM seen_subs" in the main body, then you can write:


```
if self.include_unlocatable:
  return f"{sql} RIGHT OUTER JOIN summed_weights USING(subscriber)"
return sql
```

?

Reply:
Ah, actually this'll go wrong anyway with a multicolumn spatial unit, no?

Reply:
I think so, it'd shake out to `loc_id.[lat, lon]` (which I don't think is legal SQL?)

Reply:
I'm going to add a test using `exemplar_spatial_unit_param` to catch this